File: typescript/src/terminal.ts
import * as readline from "node:readline";
import { Writable } from "node:stream";
import { ansi, BOLD, getTheme, RESET } from "./theme.js";

export async function readLine(prompt: string): Promise<string> {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stderr,
    terminal: true,
  });
  try {
    return await new Promise<string>((resolve) => {
      rl.question(prompt, (answer) => resolve(answer));
    });
  } finally {
    rl.close();
  }
}

export async function readPassword(prompt: string, mask?: string | null): Promise<string> {
  const mutableOutput = new Writable({
    write(_chunk, _encoding, callback) {
      callback();
    },
  });

  const rl = readline.createInterface({
    input: process.stdin,
    output: mutableOutput,
    terminal: true,
  });

  process.stderr.write(prompt);

  return new Promise<string>((resolve) => {
    const chars: string[] = [];
    process.stdin.setRawMode(true);
    process.stdin.resume();

    const onData = (buf: Buffer): void => {
      const ch = buf.toString("utf8");
      if (ch === "\r" || ch === "\n") {
        process.stdin.setRawMode(false);
        process.stdin.removeListener("data", onData);
        process.stderr.write("\n");
        rl.close();
        resolve(chars.join(""));
        return;
      }
      if (ch === "\x03") {
        process.stdin.setRawMode(false);
        process.stdin.removeListener("data", onData);
        rl.close();
        process.stderr.write("\n");
        resolve("");
        return;
      }
      if (ch === "\x7f" || ch === "\b") {
        if (chars.length > 0) {
          chars.pop();
          if (mask) {
            process.stderr.write("\b \b");
          }
        }
        return;
      }
      chars.push(ch);
      if (mask) {
        process.stderr.write(mask);
      }
    };
    process.stdin.on("data", onData);
  });
}

export function formatQuestion(message: string, suffix = ""): string {
  const t = getTheme();
  return `${ansi(t.question)}${t.symQuestion}${RESET} ${BOLD}${message}${suffix}:${RESET} `;
}

export function formatSuccess(message: string, answer: string): string {
  const t = getTheme();
  return `${ansi(t.success)}${t.symSuccess}${RESET} ${message} ${ansi(t.answer)}${answer}${RESET}`;
}

export function formatError(msg: string): string {
  const t = getTheme();
  return `${ansi(t.error)}  ${msg}${RESET}`;
}

export interface ListItem {
  text: string;
  style: string;
}

export interface ListConfig {
  message: string;
  getItems: () => ListItem[];
  onKey: (key: string) => { done: false } | { done: true; result: unknown };
}

export async function runListPrompt(config: ListConfig): Promise<unknown> {
  const { stdin } = process;
  if (!stdin.setRawMode) throw new Error("Terminal does not support raw mode");

  stdin.setRawMode(true);
  stdin.resume();
  stdin.setEncoding("utf8");

  let renderedLines = 0;

  const render = (): void => {
    if (renderedLines > 0) {
      process.stderr.write(`\x1b[${renderedLines}A`);
    }
    const t = getTheme();
    const header = `${ansi(t.question)}${t.symQuestion}${RESET} ${BOLD}${config.message}${RESET}`;
    const items = config.getItems();
    const lines = [header, ...items.map((it) => `${it.style}${it.text}${RESET}`)];
    for (let i = 0; i < lines.length; i++) {
      process.stderr.write(`\x1b[2K${lines[i]!}\n`);
    }
    if (renderedLines > lines.length) {
      for (let i = lines.length; i < renderedLines; i++) {
        process.stderr.write("\x1b[2K\n");
      }
      process.stderr.write(`\x1b[${renderedLines - lines.length}A`);
    }
    renderedLines = lines.length;
  };

  process.stderr.write("\x1b[?25l");
  render();

  return new Promise<unknown>((resolve) => {
    const onData = (data: Buffer | string): void => {
      const key = parseKey(typeof data === "string" ? data : data.toString("utf8"));
      const result = config.onKey(key);
      if (result.done) {
        stdin.setRawMode(false);
        stdin.removeListener("data", onData);
        stdin.pause();
        process.stderr.write("\x1b[?25h");
        if (renderedLines > 0) {
          process.stderr.write(`\x1b[${renderedLines}A`);
          for (let i = 0; i < renderedLines; i++) {
            process.stderr.write("\x1b[2K\n");
          }
          process.stderr.write(`\x1b[${renderedLines}A`);
        }
        resolve(result.result);
      } else {
        render();
      }
    };
    stdin.on("data", onData);
  });
}

function parseKey(data: string): string {
  if (data === "\x1b[A") return "up";
  if (data === "\x1b[B") return "down";
  if (data === "\r" || data === "\n") return "enter";
  if (data === " ") return "space";
  if (data === "\x03") return "ctrl-c";
  if (data === "\x1b[H" || data === "\x1b[1~") return "home";
  if (data === "\x1b[F" || data === "\x1b[4~") return "end";
  if (data.length === 1) return data;
  return "";
}

File: rust/src/terminal.rs
use crate::theme::{ansi_color, BOLD, DEFAULT_THEME, RESET};
use crossterm::cursor;
use crossterm::event::{self, Event, KeyCode, KeyEvent, KeyModifiers};
use crossterm::{queue, terminal};
use std::io::{self, BufRead, Write};

pub fn read_line(prompt: &str) -> io::Result<String> {
    read_line_with_default(prompt, None)
}

pub fn read_line_with_default(prompt: &str, prefill: Option<&str>) -> io::Result<String> {
    match prefill {
        Some(default) => {
            eprint!("{prompt}{default}");
            io::stderr().flush()?;
            let mut line = String::new();
            io::stdin().lock().read_line(&mut line)?;
            let trimmed = line
                .trim_end_matches('\n')
                .trim_end_matches('\r')
                .to_string();
            if trimmed.is_empty() {
                Ok(default.to_string())
            } else {
                Ok(trimmed)
            }
        }
        None => {
            eprint!("{prompt}");
            io::stderr().flush()?;
            let mut line = String::new();
            io::stdin().lock().read_line(&mut line)?;
            Ok(line
                .trim_end_matches('\n')
                .trim_end_matches('\r')
                .to_string())
        }
    }
}

pub fn format_question(message: &str, suffix: &str) -> String {
    let t = &DEFAULT_THEME;
    let c = ansi_color(t.question);
    format!(
        "{c}{}{RESET} {BOLD}{message}{suffix}:{RESET} ",
        t.sym_question
    )
}

pub fn format_success(message: &str, answer: &str) -> String {
    let t = &DEFAULT_THEME;
    let sc = ansi_color(t.success);
    let ac = ansi_color(t.answer);
    format!("{sc}{}{RESET} {message} {ac}{answer}{RESET}", t.sym_success)
}

pub fn format_error(msg: &str) -> String {
    let ec = ansi_color(DEFAULT_THEME.error);
    format!("{ec}  {msg}{RESET}")
}

pub enum KeyInput {
    Up,
    Down,
    Enter,
    Space,
    Backspace,
    Char(char),
    CtrlC,
    Other,
}

pub fn read_key() -> io::Result<KeyInput> {
    loop {
        if let Event::Key(KeyEvent {
            code, modifiers, ..
        }) = event::read()?
        {
            if modifiers.contains(KeyModifiers::CONTROL) && code == KeyCode::Char('c') {
                return Ok(KeyInput::CtrlC);
            }
            return Ok(match code {
                KeyCode::Up => KeyInput::Up,
                KeyCode::Down => KeyInput::Down,
                KeyCode::Enter => KeyInput::Enter,
                KeyCode::Backspace => KeyInput::Backspace,
                KeyCode::Char(' ') => KeyInput::Space,
                KeyCode::Char(c) => KeyInput::Char(c),
                _ => KeyInput::Other,
            });
        }
    }
}

/// Poll for a key event with a timeout. Returns `None` if no event is
/// available within `timeout`.
pub fn poll_key(timeout: std::time::Duration) -> io::Result<Option<KeyInput>> {
    if event::poll(timeout)? {
        Ok(Some(read_key()?))
    } else {
        Ok(None)
    }
}

#[derive(Default)]
pub struct ListRenderer {
    rendered_lines: usize,
}

impl ListRenderer {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn render(&mut self, header: &str, items: &[(String, String)]) {
        let mut stderr = io::stderr().lock();
        if self.rendered_lines > 0 {
            let _ = queue!(stderr, cursor::RestorePosition);
        }
        let _ = queue!(
            stderr,
            cursor::SavePosition,
            terminal::Clear(terminal::ClearType::FromCursorDown)
        );
        let _ = write!(stderr, "{header}");
        for (style, text) in items {
            let _ = write!(stderr, "\r\n{style}{text}{RESET}");
        }
        self.rendered_lines = 1 + items.len();
        let _ = stderr.flush();
    }

    pub fn clear(&self) {
        let mut stderr = io::stderr().lock();
        if self.rendered_lines > 0 {
            let _ = queue!(
                stderr,
                cursor::RestorePosition,
                terminal::Clear(terminal::ClearType::FromCursorDown)
            );
        }
        let _ = stderr.flush();
    }

    pub fn enable_raw() -> io::Result<()> {
        terminal::enable_raw_mode()?;
        eprint!("\x1b[?25l");
        io::stderr().flush()?;
        Ok(())
    }

    pub fn disable_raw() -> io::Result<()> {
        terminal::disable_raw_mode()?;
        eprint!("\x1b[?25h");
        io::stderr().flush()?;
        Ok(())
    }
}
